etcd/cmd/microshift-etcd/run.go

```diff
@@ -97,17 +113,67 @@ func (s *EtcdService) Run() error {
 	}
 	<-e.Server.ReadyNotify()
 
+	// Start up the defrag controller.
+	defragCtx, defragShutdown := context.WithCancel(context.Background())
+	go s.defragController(defragCtx, e.Server.Backend())
+
 	// Wait to be stopped.
 	sigTerm := make(chan os.Signal, 1)
 	signal.Notify(sigTerm, os.Interrupt, syscall.SIGTERM)
 	sig := <-sigTerm
 	klog.Infof("microshift-etcd received signal %v - stopping", sig)
 
+	// Shutdown the defrag controller.
+	defragShutdown()
+
 	e.Server.Stop()
 	<-e.Server.StopNotify()
 	return nil
 }
 
+func (s *EtcdService) defragController(ctx context.Context, be backend.Backend) error {
+	// If we were told to, go ahead and do a defragment now.
+	if doStartupDefrag {
+		if err := be.Defrag(); err != nil {
+			err = fmt.Errorf("initial defragmentation failed: %v", err)
+			klog.Error(err)
+			return err
+		}
+	}
+
+	if defragCheckFreq == 0 {
+		klog.Warning("defragmentation has been disabled")
+		return nil
+	}
+
+	// This timer will check the fragmented conditions periodically
+	timer := time.NewTimer(defragCheckFreq)
+	defer func() {
+		if !timer.Stop() {
+			<-timer.C
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			if isBackendFragmented(be) {
+				klog.Info("attempting to defragment backend")
+				start := time.Now()
+				if err := be.Defrag(); err != nil {
+					klog.Errorf("defragmentation failed: %v", err)
+					// TODO handle error
```

Comment:
In Openshift, the operator will retry a failing defrag 3 times before marking the etcd node degraded.
We could retry (configurable a default of) 3 times before throwing an error and stopping things, or maybe even allow the user to specify a value that just turns it off? Thoughts?